File: rabbitmq-streams-trading/src/batching_consumer.py
"""
Batching Consumer - High-performance micro-batching for market data.

This consumer demonstrates:
- Micro-batching for improved throughput
- Per-symbol aggregation windows
- VWAP (Volume Weighted Average Price) calculation
- Backpressure handling
"""

import asyncio
import logging
from collections import defaultdict
from datetime import datetime, timezone
from typing import Callable, List, Dict
from collections.abc import Awaitable

from rstream import (
    Consumer,
    ConsumerOffsetSpecification,
    OffsetType,
    AMQPMessage,
    MessageContext,
    amqp_decoder,
)

from .config import StreamConfig
from .models import MarketData, SymbolStats

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(name)s - %(message)s"
)
logger = logging.getLogger(__name__)

BatchHandler = Callable[[List[MarketData]], Awaitable[None]]


class BatchingConsumer:
    """High-performance batching consumer for market data."""
    
    def __init__(
        self,
        config: StreamConfig,
        batch_size: int = 1000,
        batch_timeout_ms: float = 100,
    ):
        self.config = config
        self.batch_size = batch_size
        self.batch_timeout_ms = batch_timeout_ms
        self.consumer: Consumer | None = None
        self.message_count = 0
        self.batch_count = 0
        
        self._buffer: List[MarketData] = []
        self._buffer_lock = asyncio.Lock()
        self._running = False
        self._flush_task: asyncio.Task | None = None
    
    async def start(self, handler: BatchHandler) -> None:
        """Start consuming with batching."""
        self.consumer = Consumer(
            host=self.config.host,
            port=self.config.port,
            username=self.config.username,
            password=self.config.password,
            vhost=self.config.virtual_host,
        )
        await self.consumer.start()
        
        self._running = True
        self._handler = handler
        
        # Start the periodic flush task
        self._flush_task = asyncio.create_task(self._periodic_flush())
        
        async def on_message(msg: AMQPMessage, context: MessageContext) -> None:
            try:
                data = MarketData.from_bytes(msg.body)
                self.message_count += 1
                
                async with self._buffer_lock:
                    self._buffer.append(data)
                    
                    # Flush if batch size reached
                    if len(self._buffer) >= self.batch_size:
                        await self._flush_batch()
                        
            except Exception as e:
                logger.error(f"Error buffering message: {e}")
        
        await self.consumer.subscribe(
            stream=self.config.stream_name,
            callback=on_message,
            offset_specification=ConsumerOffsetSpecification(OffsetType.FIRST, None),
            decoder=amqp_decoder,  # Decode AMQP messages
        )
        
        logger.info(
            f"BatchingConsumer started with batch_size={self.batch_size}, "
            f"timeout={self.batch_timeout_ms}ms"
        )
    
    async def _periodic_flush(self) -> None:
        """Periodically flush the buffer."""
        while self._running:
            await asyncio.sleep(self.batch_timeout_ms / 1000)
            async with self._buffer_lock:
                if self._buffer:
                    await self._flush_batch()
    
    async def _flush_batch(self) -> None:
        """Flush the current buffer as a batch."""
        if not self._buffer:
            return
        
        batch = self._buffer[:self.batch_size]
        self._buffer = self._buffer[self.batch_size:]
        
        self.batch_count += 1
        
        try:
            await self._handler(batch)
        except Exception as e:
            logger.error(f"Error processing batch: {e}")
    
    async def stop(self) -> None:
        """Stop the consumer."""
        self._running = False
        
        if self._flush_task:
            self._flush_task.cancel()
            try:
                await self._flush_task
            except asyncio.CancelledError:
                pass
        
        # Flush remaining messages
        async with self._buffer_lock:
            if self._buffer:
                await self._flush_batch()
        
        if self.consumer:
            await self.consumer.close()
        
        logger.info(
            f"BatchingConsumer closed. Messages: {self.message_count}, "
            f"Batches: {self.batch_count}"
        )
    
    @staticmethod
    def aggregate_by_symbol(batch: List[MarketData]) -> Dict[str, SymbolStats]:
        """Get aggregated statistics by symbol."""
        stats: Dict[str, SymbolStats] = {}
        
        for data in batch:
            if data.symbol not in stats:
                stats[data.symbol] = SymbolStats(data.symbol)
            stats[data.symbol].update(data)
        
        return stats


async def main():
    """Run the batching consumer."""
    config = StreamConfig.defaults()
    
    consumer = BatchingConsumer(
        config,
        batch_size=1000,
        batch_timeout_ms=100,
    )
    
    async def handle_batch(batch: List[MarketData]) -> None:
        # Aggregate statistics by symbol
        stats = BatchingConsumer.aggregate_by_symbol(batch)
        
        logger.info(f"Batch processed: {len(batch)} ticks across {len(stats)} symbols")
        
        for symbol_stats in stats.values():
            logger.info(f"  {symbol_stats}")
    
    try:
        await consumer.start(handle_batch)
        
        # Run for 60 seconds
        await asyncio.sleep(60)
        
        logger.info(
            f"Total messages: {consumer.message_count}, "
            f"batches: {consumer.batch_count}"
        )
    finally:
        await consumer.stop()


if __name__ == "__main__":
    asyncio.run(main())

File: rabbitmq-streams-trading/src/main/java/com/flowistic/trading/consumer/BatchingConsumer.java
package com.flowistic.trading.consumer;

import com.flowistic.trading.config.StreamConfig;
import com.flowistic.trading.model.MarketData;
import com.flowistic.trading.serialization.MarketDataCodec;
import com.rabbitmq.stream.Consumer;
import com.rabbitmq.stream.Environment;
import com.rabbitmq.stream.OffsetSpecification;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Duration;
import java.time.Instant;
import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicLong;

/**
 * High-performance batching consumer for market data.
 * 
 * This consumer demonstrates:
 * - Micro-batching for improved throughput
 * - Per-symbol aggregation windows
 * - VWAP (Volume Weighted Average Price) calculation
 * - Backpressure handling
 */
public class BatchingConsumer implements AutoCloseable {
    
    private static final Logger logger = LoggerFactory.getLogger(BatchingConsumer.class);
    
    private final StreamConfig config;
    private final MarketDataCodec codec;
    private final Environment environment;
    private Consumer consumer;
    
    private final BlockingQueue<MarketData> buffer;
    private final ScheduledExecutorService scheduler;
    private final ExecutorService processor;
    private final AtomicLong messageCount;
    private final AtomicLong batchCount;
    
    private final int batchSize;
    private final Duration batchTimeout;
    
    public BatchingConsumer(StreamConfig config, int batchSize, Duration batchTimeout) {
        this.config = config;
        this.batchSize = batchSize;
        this.batchTimeout = batchTimeout;
        this.codec = new MarketDataCodec();
        this.messageCount = new AtomicLong(0);
        this.batchCount = new AtomicLong(0);
        
        // Bounded buffer for backpressure
        this.buffer = new LinkedBlockingQueue<>(batchSize * 10);
        this.scheduler = Executors.newSingleThreadScheduledExecutor();
        this.processor = Executors.newFixedThreadPool(
            Runtime.getRuntime().availableProcessors()
        );
        
        this.environment = Environment.builder()
            .host(config.host())
            .port(config.port())
            .username(config.username())
            .password(config.password())
            .virtualHost(config.virtualHost())
            .build();
        
        logger.info("BatchingConsumer initialized with batchSize={}, timeout={}",
            batchSize, batchTimeout);
    }
    
    /**
     * Start consuming with batching.
     */
    public void start(BatchHandler batchHandler) {
        // Start the batch flushing timer
        scheduler.scheduleAtFixedRate(
            () -> flushBatch(batchHandler),
            batchTimeout.toMillis(),
            batchTimeout.toMillis(),
            TimeUnit.MILLISECONDS
        );
        
        // Start the consumer
        this.consumer = environment.consumerBuilder()
            .stream(config.streamName())
            .offset(OffsetSpecification.first())
            .messageHandler((context, message) -> {
                try {
                    MarketData data = codec.decode(message);
                    messageCount.incrementAndGet();
                    
                    // Try to add to buffer, apply backpressure if full
                    if (!buffer.offer(data, 100, TimeUnit.MILLISECONDS)) {
                        logger.warn("Buffer full, applying backpressure");
                    }
                    
                    // Flush if batch size reached
                    if (buffer.size() >= batchSize) {
                        flushBatch(batchHandler);
                    }
                    
                } catch (Exception e) {
                    logger.error("Error buffering message: {}", e.getMessage());
                }
            })
            .build();
        
        logger.info("BatchingConsumer started");
    }
    
    private synchronized void flushBatch(BatchHandler handler) {
        if (buffer.isEmpty()) {
            return;
        }
        
        List<MarketData> batch = new ArrayList<>(batchSize);
        buffer.drainTo(batch, batchSize);
        
        if (!batch.isEmpty()) {
            batchCount.incrementAndGet();
            processor.submit(() -> {
                try {
                    handler.handleBatch(batch);
                } catch (Exception e) {
                    logger.error("Error processing batch: {}", e.getMessage());
                }
            });
        }
    }
    
    /**
     * Get aggregated statistics by symbol.
     */
    public static Map<String, SymbolStats> aggregateBySymbol(List<MarketData> batch) {
        Map<String, SymbolStats> stats = new HashMap<>();
        
        for (MarketData data : batch) {
            stats.computeIfAbsent(data.symbol(), SymbolStats::new)
                 .update(data);
        }
        
        return stats;
    }
    
    public long getMessageCount() {
        return messageCount.get();
    }
    
    public long getBatchCount() {
        return batchCount.get();
    }
    
    @Override
    public void close() {
        scheduler.shutdown();
        processor.shutdown();
        try {
            scheduler.awaitTermination(5, TimeUnit.SECONDS);
            processor.awaitTermination(5, TimeUnit.SECONDS);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        
        if (consumer != null) {
            consumer.close();
        }
        if (environment != null) {
            environment.close();
        }
        
        logger.info("BatchingConsumer closed. Messages: {}, Batches: {}", 
            messageCount.get(), batchCount.get());
    }
    
    /**
     * Handler for processing batches of market data.
     */
    @FunctionalInterface
    public interface BatchHandler {
        void handleBatch(List<MarketData> batch);
    }
    
    /**
     * Statistics aggregation per symbol.
     */
    public static class SymbolStats {
        private final String symbol;
        private int tickCount;
        private java.math.BigDecimal sumPrice = java.math.BigDecimal.ZERO;
        private java.math.BigDecimal sumVolume = java.math.BigDecimal.ZERO;
        private java.math.BigDecimal sumPriceVolume = java.math.BigDecimal.ZERO;
        private java.math.BigDecimal high = java.math.BigDecimal.ZERO;
        private java.math.BigDecimal low = new java.math.BigDecimal("999999999");
        private java.math.BigDecimal lastPrice;
        private Instant firstTick;
        private Instant lastTick;
        
        public SymbolStats(String symbol) {
            this.symbol = symbol;
        }
        
        public void update(MarketData data) {
            tickCount++;
            var price = data.lastPrice();
            var size = data.lastSize();
            
            sumPrice = sumPrice.add(price);
            sumVolume = sumVolume.add(size);
            sumPriceVolume = sumPriceVolume.add(price.multiply(size));
            
            if (price.compareTo(high) > 0) high = price;
            if (price.compareTo(low) < 0) low = price;
            
            lastPrice = price;
            if (firstTick == null) firstTick = data.timestamp();
            lastTick = data.timestamp();
        }
        
        public java.math.BigDecimal getVwap() {
            if (sumVolume.compareTo(java.math.BigDecimal.ZERO) == 0) {
                return java.math.BigDecimal.ZERO;
            }
            return sumPriceVolume.divide(sumVolume, 4, java.math.RoundingMode.HALF_UP);
        }
        
        public java.math.BigDecimal getAvgPrice() {
            if (tickCount == 0) return java.math.BigDecimal.ZERO;
            return sumPrice.divide(java.math.BigDecimal.valueOf(tickCount), 
                4, java.math.RoundingMode.HALF_UP);
        }
        
        @Override
        public String toString() {
            return String.format(
                "SymbolStats{symbol='%s', ticks=%d, vwap=%s, avg=%s, high=%s, low=%s, last=%s}",
                symbol, tickCount, getVwap(), getAvgPrice(), high, low, lastPrice);
        }
    }
    
    public static void main(String[] args) throws Exception {
        StreamConfig config = StreamConfig.defaults();
        
        try (BatchingConsumer consumer = new BatchingConsumer(
                config, 
                1000,                           // Batch size
                Duration.ofMillis(100))) {      // Flush interval
            
            consumer.start(batch -> {
                // Aggregate statistics by symbol
                Map<String, SymbolStats> stats = aggregateBySymbol(batch);
                
                logger.info("Batch processed: {} ticks across {} symbols",
                    batch.size(), stats.size());
                
                stats.values().forEach(s -> logger.info("  {}", s));
            });
            
            // Run for 60 seconds
            Thread.sleep(Duration.ofSeconds(60).toMillis());
            
            logger.info("Total messages: {}, batches: {}", 
                consumer.getMessageCount(), consumer.getBatchCount());
        }
    }
}
